gokitmiddlewares/metricsmiddleware/middleware.go

```diff
@@ -25,23 +27,37 @@ func New(c Config) (endpoint.Middleware, error) {
 		return nil, err
 	}
 
-	shouldDecodeLabels := c.LabelsDecoder != nil
-
 	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		log.Debug().Str("config", logutil.Flatten(c)).Msg("New metrics endpoint middleware")
+
 		return func(ctx context.Context, req interface{}) (resp interface{}, err error) {
-			s := m.Begin()
-			defer func() {
-				defer func() {
-					if r := recover(); r != nil {
-						log.Ctx(ctx).Error().Msgf("Metrics middleware panicked! Please check your code and configuration: %v", r)
-					}
-				}()
-				if shouldDecodeLabels {
-					s = s.WithLabels(c.LabelsDecoder(ctx, req, resp, err))
+			log.Ctx(ctx).Debug().Msg("Start counting endpoitn mwetrics")
+
+			defer func(s metrifier.Span) {
+				var r interface{}
+				// Panics are handled as errors and re-raised
+				if r = recover(); r != nil {
+					log.Ctx(ctx).Warn().Msg("Metrics endpoint middleware is handling an uncaught a panic. Please fix it!")
+					err = errors.E(errors.Errorf("%v", r), errors.SeverityFatal, errors.CodePanic)
+				}
+				metrify(ctx, c.LabelsDecoder, s, req, resp, err)
+				if r != nil {
+					panic(r)
 				}
-				s.End(err)
-			}()
+			}(m.Begin())
 			return next(ctx, req)
 		}
 	}, nil
 }
+
+func metrify(ctx context.Context, dec LabelsDecoder, s metrifier.Span, req, resp interface{}, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			log.Ctx(ctx).Error().Msgf("Metrics middleware panicked! Please check your code and configuration: %v", r)
```

Comment:
Nit: it would be nicer if the `panic` was logged as a `Str`, leaving the `Msg` the same to all logs, making it easier to filter when inspecting logs.